### PR TITLE
Add Pythia8 UserHook for Resonance decay filtering

### PR DESCRIPTION
**Pull request by bendavid**
Allows to retry resonance decays until desired configuration is present.  This is much faster than filtering at the event level, since only the resonance decays are retried.  It also works for LHE input without throwing away any events, with the caveat that only the decays handled by pythia are retried (but decay products from the lhe level can still be included in the counting, eg for counting the combined number of leptons at LHE + pythia decay level for Higgs associated production, etc)

In order for this new userhook to work with jet matching, emission veto, etc, a further UserHook was required to act as a generic mechanism to run multiple user hooks simultaneously (and this is now used by default)

An example gen fragment is provided in 
GeneratorInterface/Pythia8Interface/test/resonanceDecayFilter.py

Immediate use case is some SUSY signal generation which requires some filtering of Higgs decays (two Higgs in each event and requesting at least one H->gamma gamma decay)

### CODE REVIEW

**Comment by bendavid** [comment]
@smrenna

**Comment by bendavid** [comment]
@covarell @govoni any chance to move this forward?

**Comment by covarell** [comment]
I was waiting for comments from @smrenna

**Comment by smrenna** [comment]
I will look at this today.  Josh and I discussed this, but I have not actually looked at the code yet.

**Comment by smrenna** [comment]
I just added a comment to the code, where I think there is one mistake.  Otherwise, it looks okay.   I may have time to try it out later today, but, otherwise, I think you should commit and we can address any issues afterwards.   Josh, do you have some validation plots or tables that demonstrate it does what we want?

**Comment by bendavid** [comment]
I checked with the JetMatchingMadgraph hook that the jet matching results don't change adding this code and combining with the resonance decay filter.  Of course that doesn't test the full range of possible UserHook functionality.

**Comment by bendavid** [comment]
And yes, agreed there is one mistake in the protections for multiple incompatible hooks, fixing now.

**Comment by bendavid** [comment]
Ok the issue with the protections is fixed now.

**Comment by covarell** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 1]
```diff
@@ -144,9 +155,9 @@ Pythia8Hadronizer::Pythia8Hadronizer(const edm::ParameterSet &params) :
   comEnergy(params.getParameter<double>("comEnergy")),
   LHEInputFileName(params.getUntrackedParameter<std::string>("LHEInputFileName","")),
   fInitialState(PP),
-  fReweightUserHook(0),fReweightRapUserHook(0),fReweightPtHatRapUserHook(0),
+  fMultiUserHook(new MultiUserHook), fReweightUserHook(0),fReweightRapUserHook(0),fReweightPtHatRapUserHook(0),
```
Comment (davidlange6) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc]: @bendavid - please delete this pointer in the destructor. Thanks

Comment (bendavid) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc]: done

**Comment by govoni** [comment]
+1

**Comment by bendavid** [comment]
please test

**Comment by davidlange6** [comment]
+1

**Pull request merged**